waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java

```diff
@@ -45,104 +38,49 @@
 public class ServerSocketRule extends ExternalResource {
   private static final Logger LOG = LoggerFactory.getLogger(ServerSocketRule.class);
 
-  private final ExecutorService executor = Executors.newSingleThreadExecutor();
-  private final AsynchronousServerSocketChannel listener;
   private final InetSocketAddress address;
   private final ByteArrayOutputStream output = new ByteArrayOutputStream();
-  private final ConcurrentLinkedQueue<Future<Void>> requests = new ConcurrentLinkedQueue<>();
+
+  private final ServerSocketChannel serverSocketChannel;
+
+  private int requests = 0;
 
   public ServerSocketRule() {
     try {
-      listener = AsynchronousServerSocketChannel.open().bind(new InetSocketAddress(0));
-      address = (InetSocketAddress) listener.getLocalAddress();
+      serverSocketChannel = (ServerSocketChannel) ServerSocketChannel
+          .open()
+          .bind(new InetSocketAddress(0))
+          .configureBlocking(false);
+      address = (InetSocketAddress) serverSocketChannel.getLocalAddress();
     } catch (IOException e) {
       throw new RuntimeException(e);
     }
   }
 
```

Comment:
With respect to all the code that's gone, any idea if it served much purpose? Are there new edge cases we might be exposed to? It seemed to be handling cases where you'd want to wait for input that had been previously been sent etc. Not sure we need it all, just curious what you thought after looking into in more detail. This looks much more straightforward and doesn't seem to exhibit the issue with reads returning data out of order that we've been seeing.

Reply:
hard to tell, there were some bugs on the sending side (where we didn't flush explicitly) so some of them were added to fix that until we fixed the flushing. 
But I am not sure of the whole history. This doesn't seem to block, requests are pending until you `accept` at which point you can read it. It's probably less performant this way but more deterministic which is fine cause we use this is meant for tests.